hack/util.sh

```diff
@@ -693,3 +693,27 @@ RUN echo -e http://mirrors.ustc.edu.cn/alpine/v3.17/main/ > /etc/apk/repositorie
   done
 }
 
+# util::wait_nodes_taint_disappear will wait for all the nodes' taint to disappear
+# Parameters:
+#  - timeout: Timeout in seconds.
+#  - kubeconfig_path: The path of kubeconfig.
+# Returns:
+#  1 if the condition is not met before the timeout, else 0
+function util::wait_nodes_taint_disappear() {
```

Comment:
I think placing the timeout inside the function would make it more user-friendly for other developers to utilize.

Reply:
cc @ikaven1024 for checking again

Reply:
Uh... the format doesn't look quite good when done this way. However, if I use tabs instead, the script will be incorrect.![image](https://user-images.githubusercontent.com/19745536/227469987-6df5a516-2501-4d1a-a76d-887592112c98.png)

Reply:
The code looks more complex with the timeout command.
The original version:
```sh
function util::wait_nodes_taint_disappear() {
  local timeout=${1}
  local kubeconfig_path=${2}
  
  while [[ $timeout > 0 ]]
  do
    taints=$(kubectl get nodes --kubeconfig=${kubeconfig_path} -o=jsonpath="{.items[*].spec.taints}")
    if [ -z $taints ]; then
      return 0
    fi
    timeout=$((timeout - 1))
    sleep 1
  done

  echo "Timeout for nodes' taint to disappear"
  return 1
}
```

The modified version:
```sh
function a() {
  local timeout=${1}
  local kubeconfig_path=${2}
  
  timeout "${timeout}" bash <<EOF
  while true
  do
    taints=$(kubectl get nodes --kubeconfig="${kubeconfig_path}" -o=jsonpath="{.items[*].spec.taints}")
    if [ -z \$taints ]; then
      exit 0
    fi
    sleep 1
  done
EOF
  result=$?

  if [[ $result != 0 ]]; then
    echo "Timeout for nodes' taint to disappear"
  fi  
  return $result
}
```

What you think? which one is better?

Reply:
I tried, it works fine
![70ce5d7903aeb369921c45a79a49e1e](https://user-images.githubusercontent.com/19745536/227889054-7a3e861e-e0b3-4a43-aa15-f37c48e31014.png)

Reply:
> PS: Your modified version has mistake, and will never not echo "Timeout for nodes' taint to disappear"

I'm not familiar with the command `timeout`, can you help to point out why it will never echo the log?

For the `original version`, my feeling is that the `timeout` essentially is a retry time, not a timeout in the usual sense. I mean the variable might be misleading.

Reply:
I talked to @jwcesign, he said there is a syntax error for https://github.com/karmada-io/karmada/pull/3323/files#r1148322580 on the line:
> timeout "${timeout}" bash <<EOF && return 0

I'm not sure if it is due to different version of timeout. @jwcesign can you share more details here?

Reply:
Sorry, my mistake, there is no syntax error for `timeout "${timeout}" bash <<EOF && return 0`. 
There is syntax error when I put return 0 as follows:
```
function util::wait_nodes_taint_disappear() {
  local timeout=${1}
  local kubeconfig_path=${2}

  timeout "${timeout}" bash <<EOF
    while true
    do
      xxx
      return 1
    done
EOF
  echo $?
  echo "Timeout for nodes' taint to disappear"
  return 1
}
```

I don't think we can use command timeout. As @ikaven1024 mentioned, with `set -o errexit;set -o nounset;set -o pipefail`, using timeout will cause the command to exit with `$? != 0` when timing out, resulting in an immediate exit.

Reply:
Thanks, I didn't know we could use a command like `timeout "${timeout}" bash <<EOF && return 0 xxx` 😳, so I changed a little bit about your version before.
Now it works fine with your original version.